extensions/amp-skimlinks/0.1/link-rewriter/link-rewriter.js

```diff
@@ -0,0 +1,242 @@
+ *   by the LinkRewriterManager.
+ */
+export class LinkRewriter {
+  /**
+   * @param {!Document|!ShadowRoot} rootNode
+   * @param {string} id
+   * @param {function(!Array<!HTMLElement>):!TwoStepsResponse} resolveUnknownLinks
+   * @param {?{linkSelector: string}=} options
+   */
+  constructor(rootNode, id, resolveUnknownLinks, options) {
+    /** @public {!../../../../src/observable.Observable} */
+    this.events = new Observable();
+
+    /** @public {string} */
+    this.id = id;
+
+    /** @private {!Document|!ShadowRoot} */
+    this.rootNode_ = rootNode;
+
+    /** @private {function(!Array<!HTMLElement>):!TwoStepsResponse} */
+    this.resolveUnknownLinks_ = resolveUnknownLinks;
+
+    /** @private {string} */
+    this.linkSelector_ = options.linkSelector || 'a';
+
+    /** @private {number} */
+    this.restoreDelay_ = 300; //ms
+
+    /** @private {!./link-replacement-cache.LinkReplacementCache} */
+    this.anchorReplacementCache_ = new LinkReplacementCache();
+  }
+
+  /**
+   * Get the replacement url associated with the anchor.
+   * @param {!HTMLElement} anchor
+   * @return {?string}
+   * @public
+   */
+  getReplacementUrl(anchor) {
+    if (!this.isWatchingLink(anchor)) {
+      return null;
+    }
+
+    return this.anchorReplacementCache_.getReplacementUrlForAnchor(anchor);
+  }
+
+  /**
+   * Get the anchor to replacement url cache.
+   * Useful to extract information for tracking purposes.
+   * @return {!AnchorReplacementList}
+   * @public
+   */
+  getAnchorReplacementList() {
+    return this.anchorReplacementCache_.getAnchorReplacementList();
+  }
+
+  /**
+   * Returns True if the LinkRewriter instance is supposed to handle
+   * this particular anchor.
+   * By default LinkRewriter handles all the links on the page but
+   * inclusion/exclusion rules can be created by using the "linkSelector"
+   * option. When using this option all the links not matching the css selector
+   * will be ignored and isWatchingLink(anchor) will return false.
+   * @param {!HTMLElement} anchor
+   * @return {boolean}
+   * @public
+   */
+  isWatchingLink(anchor) {
+    return this.anchorReplacementCache_.isInCache(anchor);
+  }
+
+  /**
+   * This function is called when the user clicks on a link.
+   * It swaps temporarly the href of an anchor by its associated
+   * replacement url but only for the time needed by the browser
+   * to handle the click on the anchor and navigate to the new url.
+   * After 300ms, if the page is still open (target="_blank" scenario),
+   * the link is restored to its initial value.
+   * @param {!HTMLElement} anchor
+   * @return {boolean} - 'true' if the linkRewriter has changed the url
+   *  'false' otherwise.
+   * @public
+   */
+  rewriteAnchorUrl(anchor) {
+    const newUrl = this.getReplacementUrl(anchor);
+    if (!newUrl || newUrl === anchor.href) {
+      return false;
+    }
+    // Save so we can restore it.
+    anchor.setAttribute(ORIGINAL_URL_ATTRIBUTE, anchor.href);
```

Comment:
But: This is not re-entrant, and it's too hard to reason about the caller to `rewriteAnchorUrl`.

Reply:
This can be called multiple times on a link in the timespan, so we will have buildup of affiliate links pointing to affiliate links pointing to the original link. Kinda like, `https://example.com/?redirect=https://example.com/?redirect=https://example.com/?redirect=original.com`